File: sdks/csharp/src/Nanoservice.Core/Middleware/MiddlewareChain.cs
using Nanoservice.Core.Node;

namespace Nanoservice.Core.Middleware;

/// <summary>
/// MiddlewareChain applies a chain of middleware to a handler.
/// </summary>
public class MiddlewareChain
{
    private readonly List<IMiddleware> _middlewares = new();

    /// <summary>
    /// Add a middleware to the chain.
    /// </summary>
    public MiddlewareChain Use(IMiddleware middleware)
    {
        _middlewares.Add(middleware);
        return this;
    }

    /// <summary>
    /// Apply all middleware to the given handler, returning the wrapped handler.
    /// Middleware is applied in order: first added wraps outermost.
    /// </summary>
    public INodeHandler Apply(INodeHandler handler)
    {
        var current = handler;
        // Apply in reverse so that the first middleware added is the outermost wrapper
        for (int i = _middlewares.Count - 1; i >= 0; i--)
        {
            current = _middlewares[i].Wrap(current);
        }
        return current;
    }
}

File: sdks/java/src/main/java/com/blok/blok/middleware/MiddlewareChain.java
package com.blok.blok.middleware;

import com.blok.blok.node.NodeHandler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Composes multiple middleware into a single middleware.
 * Middleware is applied in order: first middleware is outermost.
 */
public class MiddlewareChain implements Middleware {

    private final List<Middleware> middlewares;

    /**
     * Creates a chain from the given middlewares.
     *
     * @param middlewares the middlewares to chain
     */
    public MiddlewareChain(Middleware... middlewares) {
        this.middlewares = new ArrayList<>();
        if (middlewares != null) {
            this.middlewares.addAll(Arrays.asList(middlewares));
        }
    }

    /**
     * Creates a chain from a list of middlewares.
     *
     * @param middlewares the middlewares to chain
     */
    public MiddlewareChain(List<Middleware> middlewares) {
        this.middlewares = middlewares != null ? new ArrayList<>(middlewares) : new ArrayList<>();
    }

    /**
     * Adds a middleware to the end of the chain.
     *
     * @param middleware the middleware to add
     * @return this chain for fluent use
     */
    public MiddlewareChain add(Middleware middleware) {
        if (middleware != null) {
            middlewares.add(middleware);
        }
        return this;
    }

    @Override
    public NodeHandler wrap(NodeHandler handler) {
        NodeHandler current = handler;
        for (int i = middlewares.size() - 1; i >= 0; i--) {
            current = middlewares.get(i).wrap(current);
        }
        return current;
    }
}
